storage/sql/coresql/wrapper/wrapper.go

```diff
@@ -0,0 +1,110 @@
+	"database/sql"
+
+	"github.com/google/trillian/storage"
+)
+
+// TreeWrapper provides SQL wrapping for raw tree storage.
+type TreeWrapper interface {
+	GetTreeRevisionIncludingSize(tx *sql.Tx, treeID, treeSize int64) (int64, int64, error)
+	GetSubtrees(tx *sql.Tx, treeID, treeRevision int64, nodeIDs []storage.NodeID, subtreeScanFn func(*sql.Rows) error) error
+	// SetSubtrees args should be a 4 tuple of (treeID, prefix, subtreeBytes, writeRevision) for each new subtree
+	SetSubtrees(tx *sql.Tx, args []interface{}) error
+}
+
+// LogStatementProvider provides SQL statement objects for log storage.
+type LogStatementProvider interface {
+	GetActiveLogsStmt(tx *sql.Tx) (*sql.Stmt, error)
+	GetActiveLogsWithWorkStmt(tx *sql.Tx) (*sql.Stmt, error)
+	DeleteUnsequencedStmt(tx *sql.Tx, num int) (*sql.Stmt, error)
+	GetLeavesByIndexStmt(tx *sql.Tx, num int) (*sql.Stmt, error)
+	GetLeavesByMerkleHashStmt(tx *sql.Tx, num int, orderBySequence bool) (*sql.Stmt, error)
+	GetLeavesByLeafIdentityHashStmt(tx *sql.Tx, num int) (*sql.Stmt, error)
+	InsertTreeHeadStmt(tx *sql.Tx) (*sql.Stmt, error)
+	GetLatestSignedLogRootStmt(tx *sql.Tx) (*sql.Stmt, error)
+	GetQueuedLeavesStmt(tx *sql.Tx) (*sql.Stmt, error)
+	InsertUnsequencedEntryStmt(tx *sql.Tx) (*sql.Stmt, error)
+	InsertUnsequencedLeafStmt(tx *sql.Tx) (*sql.Stmt, error)
+	InsertSequencedLeafStmt(tx *sql.Tx) (*sql.Stmt, error)
+	GetSequencedLeafCountStmt(tx *sql.Tx) (*sql.Stmt, error)
+}
+
+// MapStatementProvider provides SQL statement objects for map storage.
+type MapStatementProvider interface {
+	GetMapLeafStmt(tx *sql.Tx, num int) (*sql.Stmt, error)
+	GetLatestMapRootStmt(tx *sql.Tx) (*sql.Stmt, error)
+	InsertMapHeadStmt(tx *sql.Tx) (*sql.Stmt, error)
+	InsertMapLeafStmt(tx *sql.Tx) (*sql.Stmt, error)
+}
+
+// AdminStatementProvider provides SQL statement objects for administration.
+type AdminStatementProvider interface {
+	GetTreeIDsStmt(tx *sql.Tx) (*sql.Stmt, error)
+	GetAllTreesStmt(tx *sql.Tx) (*sql.Stmt, error)
+	GetTreeStmt(tx *sql.Tx) (*sql.Stmt, error)
+	InsertTreeStmt(tx *sql.Tx) (*sql.Stmt, error)
+	InsertTreeControlStmt(tx *sql.Tx) (*sql.Stmt, error)
+	UpdateTreeStmt(tx *sql.Tx) (*sql.Stmt, error)
+}
+
+// CustomBehaviourProvider abstracts database specific features, for example error code checking.
+type CustomBehaviourProvider interface {
+	CheckDatabaseAccessible(ctx context.Context) error
+	IsDuplicateErr(err error) bool
+	TreeRowExists(treeID int64) error
+}
+
+// LifecycleHooks allows implementations to add custom logic at various points in the
+// database and transaction flow.
+type LifecycleHooks interface {
+	OnOpenDB() error
+}
+
+// DBWrapper encapsulates a database and provides customized SQL statement objects for all types
+// of tree storage as well as abstracting some operations that differ between databases.
+// Statements returned by any of public functions belong to the caller transaction and must be
+// closed on completion of the work.
+type DBWrapper interface {
```

Comment:
Yes I'm well aware of all that. Let's try to address some of these points.

This PR didn't start out as a mix of statement provider and DAO and won't be once I finish the requested rework so that it all looks like the tree interface.

The design goals were for performance and abstraction at the storage level, which I think have been achieved. "dumb DAOs" may be OK for leaf entities but did not perform well enough for subtrees in early prototyping so we were pushed towards a more complex solution.

The mixture of layers is a fair point but I don't agree that supporting multiple databases using this approach is complex. It needs a different wrapper but it shares most of the code that drives it.  I may be able to fully encapsulate the database access within the wrapper, which will help with testing the common code. But you can't test implementations that exercise specific databases with mocks with any real confidence.

If you look at the early history of this PR I did start off with something that just plugged queries into entirely code but @AlCutter asked me to move away from that. Several other options were unworkable or had undesirable side effects.

We have a fairly immediate requirement to support Postgres, which is what I'm trying to address at the moment without making a copy of all the code. We can always revisit the overall design again later.